### PR TITLE
auto restore snapshots from zip on cold start

### LINKED ISSUE #192: Automatic Cluster Snapshot loading on Cold Start
# Automatic table loading from available Snapshots on Cold Start

When a node is started as first node in an erlang cluster (or when unable to join any pre-existing cluster), here after referred to as “cold start”, IMEM should be able to automatically load last known good configuration (to be referred to as “cluster snapshot”) and be operational and ready for handling traffic quickly without requiring much operator assistance.
1.	Loading of cluster snapshot should be automatic (if enabed) on cold start of a node.
2.	All nodes should be able to detect a normal start/restart (not cold start) and NOT attempt to load cluster snapshot even if auto loading of cluster snapshot is enabled

# Background

IMEM has the ability of periodically copy a configured set of tables (containing important / useful information / configuration for the operation of the software) and dump into disk on EACH of the cluster. This feature is referred to as “Cluster Snapshot”.

A cluster snapshot consists of .zip containing dumps of a set of table contents as .bkp files. This feature is controlled through the following three configuration parameters in ddConfig table:
•	snapshotCluster : Enable / Disable Cluster snapshot
•	snapshotClusterTables : Set of tables to snapshot abd archieve on each noe of the cluster
•	snapshotClusterHourOfDay : Hour of day (in 24 hours format) of of making such a snapshot each day

When “Cluster Snapshot” is configured and enabled each node of cluster will create a .zip cluster snapshot file at different hours of the day (as configured).

## Manual cluster recovery with cluster snapshot on cold start

In a situation when a “cold start” required due to a cluster crash (if happens). File creation timestamp of such cluster snapshot archieves can be comparied to determine the node to start cold start of cluster with which has the latest cluster snapshot archive and restore the necessary tables to arrive at a fully configured node with last known good configuration.

Once the fist node of the cluster is restored from cluster snapshot other nodes of the cluster can be started one-after-another, they will automatically arrive into directly fully functional state as the join the cluster and receive last known good configuration from cluster IMEM Database.

Once all nodes in the cluster are started, the first node (which initiated the cold start by restoring cluster snapshot), MUST be restarted to ensure a full synchronized state of the cluster.

## Proposed Solution

This solution will make the above cold start and restore of the first node, automatic. The benefit of a automatic cold start is ease of operation and quick recovery of cluster after a cluster failure.
The following points must be noted to make sure that automatic cold start is possible:
* cluster snapshot is enabled and configured as described above
* different nodes of the cluster should have different cluster snapshot hours evenly distributed over the day

If the above conditions are met then steps to automatic cold start after a total cluster loss due of site failure (or any other reason) will be as follows:
1. Determine the which node in cluster has the latest cluster snapshot archived
1. Start the node from Step 1

Automatic cold start DB restore will function as follows
1. If `cold_start_recover` application flag is `true` proceed with steps below
1. If current node is **NOT** the only node of the cluster **skip cluster automatic restore with a message**
1. Find latest cluster snapshot .zip file (from all snapshot/*.zip), if none is found **skip cluster automatic restore with a message**
1. Restore all tables from this zip

### PR DESCRIPTION
**Pull request by shamis**
fixes  #192

### CODE REVIEW

**Comment by shamis** [comment]
@c-bik feature is implemented as per the requirement, please verify

File: src/imem_snap.erl [hunk 1]
```diff
@@ -828,3 +832,14 @@ exclude_table_pattern(TablePattern) when is_list(TablePattern) ->
     ExPatterns = ?GET_SNAPSHOT_EXCLUSION_PATTERNS,
     Remark = list_to_binary(["Added ", TablePattern, " at ", imem_datatype:timestamp_to_io(?TIMESTAMP)]),
     ?PUT_SNAPSHOT_EXCLUSION_PATTERNS(lists:usort([TablePattern | ExPatterns]), Remark).
+
+restore_tables_from_latest_zip(SnapDir) ->
+    case imem_meta:nodes() of 
+        [] ->
+            case lists:reverse(lists:sort(filelib:wildcard(?BKP_ZIP_PREFIX ++ "*.zip", SnapDir))) of
```
Comment (c-bik) [inline src/imem_snap.erl]: `++` isn't necessary when using macro

File: src/imem_snap.erl [hunk 2]
```diff
@@ -828,3 +832,14 @@ exclude_table_pattern(TablePattern) when is_list(TablePattern) ->
...
+                [] -> ?Warn("No zip file found to restore tables from in ~p", [SnapDir]);
+                [ZipFile | _ ] -> restore(zip, filename:join(SnapDir, ZipFile), [], replace, false)
+            end;
+        _ -> ?Info("Other nodes exist in the cluster so not restoring tables from snapshot")
+    end.
```
Comment (c-bik) [inline src/imem_snap.erl]: Different logs:
1. if `cold_start_recover == false andalso imem_meta:nodes() == []` :
    `?Warn("Cold Start : auto restore from cluster snapshot is disabled")`
1. if `cold_start_recover == true andalso imem_meta:nodes() /= []` :
    `?Info("Not Cold Start : auto restore from cluster snapshot is skipped")`
1. if `cold_start_recover == true andalso imem_meta:nodes() == [] andalso no ZipFile found` :
    `?Warn("Cold Start : unable to auto restore, no "?BKP_ZIP_PREFIX"*.zip found in snapshot directory ~s", [SnapDir]])`
1. if `cold_start_recover == true andalso imem_meta:nodes() == [] andalso ZipFile found` :
    `?Info("Cold Start : auto restoring "?BKP_ZIP_PREFIX"*.zip found at ~s", [SnapDir]])`

**Comment by shamis** [comment]
@c-bik please verify, review comments incorporated

File: src/imem_snap.erl [hunk 3]
```diff
@@ -828,3 +828,19 @@ exclude_table_pattern(TablePattern) when is_list(TablePattern) ->
...
+                [] ->
+                    case lists:reverse(lists:sort(filelib:wildcard(?BKP_ZIP_PREFIX"*.zip", SnapDir))) of
+                        [] -> ?Warn("Cold Start : unable to auto restore, no "?BKP_ZIP_PREFIX"*.zip found in snapshot directory ~s", [SnapDir]);
+                        [ZipFile | _ ] ->
+                            ?Info("Cold Start : auto restoring ~s found at ~s", [ZipFile, SnapDir]),
+                            restore(zip, filename:join(SnapDir, ZipFile), [], replace, false)
+                    end;
+                _ -> ?Info("Not Cold Start : auto restore from cluster snapshot is skipped")
+            end;
+        _ -> ok
```
Comment (c-bik) [inline src/imem_snap.erl]: Missing first log from https://github.com/K2InformaticsGmbH/imem/pull/194#discussion_r174858543

Merge the first two nested `case` clauses into a singe case clause:
```erlang
case {application:get_env(imem, cold_start_recover), imem_meta:nodes()} of
    {{ok, true}, []} ->
...
end
```

**Pull request merged**